### PR TITLE
Add helper function to fetch constant buffer data for D3D12 debugging

### PR DESCRIPTION
**Pull request by skarolewics**

### CODE REVIEW

File: renderdoc/driver/d3d12/d3d12_shaderdebug.cpp [hunk 1]
```diff
@@ -324,6 +324,91 @@ void D3D12DebugManager::CreateShaderGlobalState(ShaderDebug::GlobalState &global
   global.PopulateGroupshared(dxbc->GetDXBCByteCode());
 }
 
+void GatherConstantBuffers(WrappedID3D12Device *pDevice, DXBC::ShaderType shaderType,
+                           const D3D12RenderState::RootSignature &rootsig,
+                           bytebuf cbufData[D3D12_COMMONSHADER_CONSTANT_BUFFER_API_SLOT_COUNT])
```
Comment (baldurk) [inline renderdoc/driver/d3d12/d3d12_shaderdebug.cpp:329]: I don't understand why `D3D12_COMMONSHADER_CONSTANT_BUFFER_API_SLOT_COUNT` exists, I think a lot of those constants were copy-pasted from D3D11 and renamed, given that `D3D12_SHADER_MAX_INTERFACES` exists but D3D12 doesn't support shader interfaces at all...

Either way this limit only applies to 'legacy' SM5.0 shaders that are limited to D3D11 compatibility. SM5.1 shaders can have any register value. E.g.:

```
cbuffer b : register(b100000000)
{
  float4 a;
};

float4 main() : SV_Target0
{
  return a;
}
```

produces

```
ps_5_1
dcl_globalFlags refactoringAllowed
dcl_constantbuffer CB0[100000000:100000000][1], immediateIndexed, space=0
dcl_output o0.xyzw
mov o0.xyzw, CB0[100000000][0].xyzw
ret
```

This also doesn't account for register spaces, which can also be arbitrary numbers and can allow the same register index to exist in multiple space with different cbuffers.

Bottom line is that this will have to be a map from `{space, register}` to data.

**Review by baldurk — CHANGES_REQUESTED** [review]
You won't be able to get away with a simple array in D3D12 unfortunately, it'll have to be a map given the sparse binding points.

This also brings up the same issue as a previous PR, that sooner rather than later we'll probably need to defer fetching most of these cbuffers as someone will come along with GBs of total cbuffer data available to the shader :(.

**Comment by skarolewics** [comment]
I agree that these changes need to happen. I'm not sure yet what downstream changes would need to occur to support this properly. I was planning to do another pass on a lot of the root signature traversal logic after getting more shader debug changes PR'd.

I think for constant buffers, it'd be a relatively light change to support, if you'd prefer that I do that now. Though I anticipate a more holistic change that will need to happen similarly for SRVs, UAVs, etc.

**Comment by baldurk** [comment]
Fair point that the SRVs/UAVs have a similar issue, I should have caught that in an earlier review which is my bad.

I think I'm OK merging this for now with a plan to fix later if that's how you want to do it, but it must be fixed before hooking up `DebugVertex/Pixel/Thread` as this would crash as soon as someone debugs a shader with a register index out of bounds.

I guess you could have a version check in those functions which bails out when it detects SM5.1, but I don't know how many people writing D3D12 code build for SM5.0 or lower when SM5.1 is a guaranteed minspec? Is a SM5.0 only debugger worthwhile?

Indeed there will need to be changes to the register lookups in the debugging code as well, especially to handle indirect access to resources (previously the resource indices were all always compile-time fixed).

**Comment by skarolewics** [comment]
Yes, there's plenty of work that still needs to happen before the UI can allow any D3D12 debugging. This is just another thing on the pile. :) Would you prefer that I create an issue to track it, or is that too specific for the issue tracker?

I don't see much benefit to handling SM5.0 differently, and in fact it may become a detriment if there are then RenderDoc bugs specific to a compiler version. And there's plenty to support with SM6.0+ that is far more deserving.

**Comment by baldurk** [comment]
No need to make an issue for it, as long as you're aware of it and it gets fixed that's good enough for me.

About SM5.0, I don't mean treating it differently necessarily I just meant that with SM5.0 then because the shader is D3D11 compatible there are no register spaces, and SRV registers are limited to 128 and cbuffer registers limited to 14, so the code as it is would work without change. What I'm not sure of is whether "shader debugging, but only for SM5.0" is a worthwhile feature on its own or whether. If it were, there could be value in targetting SM5.0 debugging first as a useful feature and then subsequently incrementally expand to SM5.1 which we can more-or-less treat as a superset of SM5.0.

SM6.0 is a whole separate issue because of DXIL and is a complete & utter mess :(. Fortunately there's very little overlap so we don't have to worry about it with any of this stuff.
